src/SpendPermissionManager.sol

```diff
@@ -445,12 +449,18 @@ contract SpendPermissionManager is EIP712 {
             revert SpendValueWithdrawAmountMismatch(value, withdrawRequest.amount);
         }
 
-        // Extract spender address from nonce (rightmost 160 bits)
-        address encodedSpender = address(uint160(withdrawRequest.nonce));
+        // Get the hash of the spend permission
+        bytes32 permissionHash = getHash(spendPermission);
+        
+        // Take first NONCE_HASH_BITS bits of the hash
+        uint256 expectedHashPortion = uint256(permissionHash) >> (256 - NONCE_HASH_BITS);
+        
+        // Extract hash portion from nonce (leftmost NONCE_HASH_BITS bits)
+        uint256 encodedHashPortion = withdrawRequest.nonce >> (256 - NONCE_HASH_BITS);
         
-        // Verify encoded spender matches actual spender
-        if (encodedSpender != spendPermission.spender) {
-            revert InvalidWithdrawRequestSpender(encodedSpender, spendPermission.spender);
+        // Verify encoded hash portion matches expected
+        if (encodedHashPortion != expectedHashPortion) {
```

Comment:
manual bit shifts feels less readable than just doing comparison on casted `bytes16`? This seems to save lines and improve readability, what do you think?
```suggestion
        if (bytes16(permissionHash) != bytes16(bytes32(withdrawRequest.nonce))) {
```

Reply:
yeah I like it! this puts the hash bits as the lower bits of the nonce, which I thinks makes potentially even more sense than the upper bits since it simplifies the extraction. casting approach ended up looking slightly different from this but I still think it's an improvement